src/monitor/notifications.c

```diff
@@ -74,22 +74,27 @@ NotifyStateChange(AutoFailoverNode *node, char *description)
 	 */
 	eventid = InsertEvent(node, description);
 
-	/*
-	 * Rather than try and escape dots and colon characters from the user
-	 * provided strings formationId and nodeHost, we include the length of the
-	 * string in the message. Parsing is then easier on the receiving side too.
-	 */
-	appendStringInfo(payload,
-					 "S:%s:%s:%lu.%s:%d:%d:%lu.%s:%d",
-					 ReplicationStateGetName(node->reportedState),
-					 ReplicationStateGetName(node->goalState),
-					 strlen(node->formationId),
-					 node->formationId,
-					 node->groupId,
-					 node->nodeId,
-					 strlen(node->nodeHost),
-					 node->nodeHost,
-					 node->nodePort);
+	/* build a json object from the notification pieces */
+	appendStringInfoChar(payload, '{');
+
+	appendStringInfo(payload, "\"type\": \"state\", ");
+
+	appendStringInfo(payload, "\"formation\": \"%s\", ",
+					 node->formationId);
+
+	appendStringInfo(payload, "\"groupId\": %d, ", node->groupId);
+	appendStringInfo(payload, "\"nodeId\": %d, ", node->nodeId);
+	appendStringInfo(payload, "\"name\": \"%s\", ", node->nodeName);
+	appendStringInfo(payload, "\"host\": \"%s\", ", node->nodeHost);
```

Comment:
This will result in invalid JSON if nodeName has `"`. Doesn't postgres have some easy to use json formatting/quoting utilities, because of JSONB? If not, why did we move away from the previous format? 

I guess this edgecase shouldn't really pop up normally, since names are unlikely to have `"` characters in them. But you never know with users, somehow they always end up doing something that breaks your stuff.

Reply:
That's a good point. I looked at mimicking more of `json_build_object` as in https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/json.c#L991  ; and then I saw other code in Postgres where they just compose the JSON entries like I did here. We could also copy paste the following code in our extension and compose the JSON notification from those pieces?

``` c
/*
 * Append JSON text for "val" to "result".
 *
 * This is just a thin wrapper around datum_to_json.  If the same type will be
 * printed many times, avoid using this; better to do the json_categorize_type
 * lookups only once.
 */
static void
add_json(Datum val, bool is_null, StringInfo result,
		 Oid val_type, bool key_scalar)
{
	JsonTypeCategory tcategory;
	Oid			outfuncoid;

	if (val_type == InvalidOid)
		ereport(ERROR,
				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
				 errmsg("could not determine input data type")));

	if (is_null)
	{
		tcategory = JSONTYPE_NULL;
		outfuncoid = InvalidOid;
	}
	else
		json_categorize_type(val_type,
							 &tcategory, &outfuncoid);

	datum_to_json(val, is_null, result, tcategory, outfuncoid, key_scalar);
}
```

Reply:
Turns out that Postgres exposes this escaping facility that we can use directly, now using it:

``` c
	escape_json(payload, node->formationId);
```